### ISSUE TITLE
Risk with composer update with unpushed feature branches

### ISSUE THREAD

**Issue by tractorcow** [composer/composer#5572]
As a result of change in behaviour from the following two changes: https://github.com/composer/composer/pull/4878 and https://github.com/composer/composer/pull/3633, it now becomes a lot risker to perform `composer update` within repositories which have installed dependencies with unpushed changes.

My project has the root composer.json as below:

```json
{
	"name": "silverstripe/installer",
	"description": "The SilverStripe Framework Installer",
	"require": {
		"php": ">=5.3.2",
		"silverstripe/cms": "3.1.x-dev",
		"silverstripe/framework": "3.1.x-dev",
		"silverstripe-themes/simple": "*"
	},
	"require-dev": {
		"phpunit/PHPUnit": "~3.7@stable"
	},
	"config": {
		"process-timeout": 600
	},
	"minimum-stability": "dev"
}
```

Typically I am working on changes within the silverstripe/framework repo, and prior to the above changes, I could reliably perform composer update periodically, and have any other module I have installed automatically updapte (as well as the framework module). I normally develop my changes on feature branches which don't typically live upstream.

However, the result of #3633 is that having unpushed changes (on these branches) results in an error, rather than silently just doing the update, which before would switch back to the original branch and pull down upstream changes.

Although #4878 catches this error, it offers two solutions which do not match either of the desired outcome.

The result of running `composer update` on my project looks similar to the below output:

```
  - Updating silverstripe/framework (3.1.x-dev 20d1167 => 3.1.x-dev 9621392)
    Update failed (Source directory framework/ has unpushed changes on the current branch: 
Branch pulls/3.1/fix-select-escaping could not be found on the origin remote and appears to be unpushed)
    Would you like to try reinstalling the package instead [yes]? 
```

If I select yes (or hit return) I have my repo deleted, including my unpushed pulls/3.1/fix-select-escaping branch (which of course, would be my own dumb fault for selecting the option, or for not pushing my changes to my fork!)

However, the alternative to this is that installation halts.

The legacy behaviour, which is the one I would prefer to retain, is that framework would swap to whichever branch / tag it wished to update to, and would not delete any other branch which contained my local changes. I could easily then switch back to that branch to continue my work. In fact, if there were no changes upstream, it would not switch the branch at all.

Could I suggest an enhancement to #3633 which would not raise the error if the current branch would otherwise be untouched? It feels like this is unnecessarily an error which could simply be logged to the stderr without interrupting the update.

Of course, if I were on 3.1 branch with unpushed changes, and composer were attempting to update to upstream changes on the same branch, it would certainly be an error. This seems to be the use case that #3633 is attempting to address?

**Comment by tractorcow** [comment]
Thanks @alcohol , I concur on the title and assignment. :)

**Comment by Seldaek** [comment]
I agree with your proposal, if the update can be done gracefully, we should only warn and ask if we should update or cancel, rather than reinstall or cancel.

**Comment by fgm** [comment]
Has this behaviour changed since that issue is almost 2 years old ?

**Comment by alcohol** [comment]
Most likely it has not changed.

**Comment by cezary-zeglen** [comment]
I would love to see this introduced, the current behaviour is quite irritating

**Comment by alcohol** [comment]
Unfortunately there is far too much work on the backlog and not enough time available. We always welcome contributions :-)

**Comment by rchavik** [comment]
I'm going to try to merge this

**Comment by rchavik** [comment]
> I'm going to try to merge this

Sorry, commenting on wrong issue :)

**Comment by Elaine-dev** [comment]
What is the workaround for this? I can't run composer remove....
